### PR TITLE
Disallow access from uncensored countries

### PR DESCRIPTION
**Pull request by oxtoacart**
@aranhoide This is ready for review.

### CODE REVIEW

File: src/github.com/getlantern/geolookup/geolookup.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
+package geolookup
```
Comment (oxtoacart) [inline src/github.com/getlantern/geolookup/geolookup.go:1]: Existing library.

File: src/github.com/hashicorp/golang-lru/.gitignore [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Compiled Object files, Static and Dynamic libs (Shared Objects)
```
Comment (oxtoacart) [inline src/github.com/hashicorp/golang-lru/.gitignore:1]: 3rd party library for LRU cache.

**Comment by oxtoacart** [comment]
@myleshorton If you wanted to, you could take a gander at this too.

File: src/github.com/getlantern/flashlight/flashlight.go [hunk 3]
```diff
@@ -138,6 +138,10 @@ func runServerProxy(cfg *config.Config) {
 			ServerCertFile: config.InConfigDir("servercert.pem"),
 		},
 		AllowedPorts: []int{80, 443, 8080, 8443, 5222, 5223, 5228},
+
+		// We allow all censored countries plus us, es and mx because we do work
+		// and testing from those countries.
+		AllowedCountries: []string{"US", "ES", "MX", "CN", "VN", "IN", "IQ", "IR", "CU", "SY", "SA", "BH", "ET", "ER", "UZ", "TM", "PK", "TR", "VE"},
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/flashlight.go:144]: List of censored countries taken from here - https://github.com/getlantern/lantern/blob/devel/src/main/java/org/lantern/DefaultCensored.java#L33

**Comment by aranhoide** [comment]
I'll take a look at this now.  Sorry I didn't get to it before; daughter is with fever.

First off, have you noticed the travis failure?

File: src/github.com/getlantern/flashlight/server/server.go [hunk 4]
```diff
@@ -120,7 +125,30 @@ func (server *Server) ListenAndServe() error {
...
-		AllowedPorts:               server.AllowedPorts,
+	}
+
+	if server.AllowedCountries != nil {
+		server.geoCache, _ = lru.New(1000000)
+	}
+
+	if server.AllowedPorts != nil || server.AllowedCountries != nil {
+		fs.Allow = func(req *http.Request, destAddr string) error {
+			if server.AllowedPorts != nil {
```
Comment (aranhoide) [inline src/github.com/getlantern/flashlight/server/server.go:136]: I like that the cheapest check is performed first.

File: src/github.com/getlantern/flashlight/server/server.go [hunk 5]
```diff
@@ -222,6 +250,69 @@ func (server *Server) stopNattywad() {
 	server.waddellClient = nil
 }
 
+func (server *Server) checkForDisallowedPort(addr string) error {
+	_, portString, err := net.SplitHostPort(addr)
```
Comment (aranhoide) [inline src/github.com/getlantern/flashlight/server/server.go:254]: Just checking: we never ever get just the host part and assume the port to be 80, right?

File: src/github.com/getlantern/flashlight/server/server.go [hunk 6]
```diff
@@ -222,6 +250,69 @@ func (server *Server) stopNattywad() {
...
+
+	country := ""
+	cachedCountry, found := server.geoCache.Get(clientIp)
+	if found {
+		log.Tracef("Country for %v found in cache", clientIp)
+		country = cachedCountry.(string)
+	} else {
+		log.Tracef("Country for %v not cached, perform geolookup", clientIp)
+		city, err := geolookup.LookupCity(clientIp)
+		if err != nil {
```
Comment (aranhoide) [inline src/github.com/getlantern/flashlight/server/server.go:294]: If it's worth caching the result of successful lookups, why is it not worth caching failures?

File: src/github.com/getlantern/flashlight/flashlight.go [hunk 7]
```diff
@@ -138,6 +138,10 @@ func runServerProxy(cfg *config.Config) {
 			ServerCertFile: config.InConfigDir("servercert.pem"),
 		},
 		AllowedPorts: []int{80, 443, 8080, 8443, 5222, 5223, 5228},
+
+		// We allow all censored countries plus us, es and mx because we do work
```
Comment (aranhoide) [inline src/github.com/getlantern/flashlight/flashlight.go:142]: ~~Super-nit: Countries are normalized to uppercase so I'd spell them that way in this comment to reinforce the point.~~ Nevermind; I see they're actually normalized at check time.

**Comment by aranhoide** [comment]
LGTM; QAing...

**Comment by aranhoide** [comment]
`testandcover.bash` worked alright.

To test the geoip filtering, I deployed this branch to a fl-nl-ara-003 and accessed it with the flashlight.yaml below (I used the same binary for the client side).  It worked from my home (which is OK), [EDIT:] and I got correctly filtered out when trying to use it from my cloudmaster in Amsterdam.

So the only thing bothering me a bit is the travis failure.  The error is testing fdcount, which this PR doesn't touch, so I suppose it's a one-time hiccup.  I'll push a noneffectual commit and see if that fixes it.

```
version: 0
cloudconfig: ""
cloudconfigca: ""
addr: localhost:9999
role: client
instanceid: ""
country: xx
statsaddr: ""
cpuprofile: ""
memprofile: ""
waddellcert: ""
stats:
  reportingperiod: "0"
  statshubaddr: pure-journey-3547.herokuapp.com
server:
  unencrypted: false
  country: ""
  registerat: ""
  portmap: 0
  advertisedhost: ""
  waddelladdr: ""
client:
  minqos: 0
  dumpheaders: false
  frontedservers:
  - host: fl-nl-ara-003.getiantem.org
    port: 443
    poolsize: 30
    masqueradeset: cloudflare
    maxmasquerades: 20
    insecureskipverify: false
    bufferrequests: false
    dialtimeoutmillis: 0
    redialattempts: 2
    weight: 1000
    qos: 2
  chainedservers: {}
  peers: {}
  masqueradesets:
    cloudflare:
    - domain: 10minutemail.com
      ipaddress: 162.159.250.16
    - domain: 1news.az
# [More masquerades and CA certs elided]
```

**Comment by aranhoide** [comment]
Yeah, it was a travis hiccup.  I'll deploy this.

File: src/github.com/getlantern/flashlight/server/server.go (hunk 5 above)
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/server/server.go:254]: Correct.

File: src/github.com/getlantern/flashlight/server/server.go (hunk 6 above)
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/server/server.go:294]: An error in this case means a network error calling the geoip service.  If the geoip service returns an unknown country, that's not an error, just a blank country.

**Pull request merged**